src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManagerBase.cs

```diff
@@ -267,6 +288,83 @@ public override Task RemoveFromGroupAsync(string connectionId, string groupName,
+            }
+
+            if (IsInvalidArgument(methodName))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(methodName));
+            }
+            // globally distinct invocationId
+            // $"{connectionId}{_callerId}{_clientResults.GetNewInvocation()}";
+            var invocationId = _clientInvocationManager.Caller.GenerateInvocationId(connectionId);
+            
+            try
+            {
+                var message = AppendMessageTracingId(new ClientInvocationMessage(invocationId, connectionId, _callerId, SerializeAllProtocols(methodName, args, invocationId)));
+                await WriteAsync(message);
+                if (_clientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+                {
+                    var instanceId = clientConnectionContext.InstanceId;
+                    var task = _clientInvocationManager.Caller.AddInvocation<T>(connectionId, invocationId, instanceId, cancellationToken);
+                    return await task;
+                }
+                else
+                {
+                    throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+                }
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+
+        // Only route server will reach here
+        public override async Task SetConnectionResultAsync(string connectionId, CompletionMessage result)
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+            if (_clientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+            {
+                // Current server is a route server.
+                // In order to inform original Caller server with the completion result, send a ClientCompletionMessage to service and the service will route ClientCompletionMessage to the original Caller server.
+                if (_clientInvocationManager.Router.ContainsInvocation(result.InvocationId))
+                {
+                    var protocol = clientConnectionContext.Protocol;
+                    var message = AppendMessageTracingId(new ClientCompletionMessage(result.InvocationId, connectionId, _callerId, protocol, SerializeCompletionMessage(result, protocol)));
+                    await WriteAsync(message);
+
+                    _clientInvocationManager.Router.TryCompleteResult(connectionId, result);
+                }
+                else
+                // Current server is the original Caller server. Complete the corresponding client invocation locally.
+                {
+                    _clientInvocationManager.Caller.TryCompleteResult(connectionId, result);
```

Comment:
Still need to send back service, as service will rely 'ClientCompeletionMessage` to do cleanup.